Add runApplication() Kotlin top level function

sdeleuze opened the pull request:
Since Kotlin extensions do not apply to static methods, this commit introduces a `runApplication()` Kotlin top level function that acts as a Kotlin shortcut for `SpringApplication.run()`.

This shortcut avoids to require using non-idiomatic code like `SpringApplication.run(FooApplication::class.java)` and provides a `runApplication<FooApplication>()` alternative (as well as an array of `KClass` based alternative when multiple classes need to be passed as parameter).

This commit also adds Kotlin main artifacts to Spring Boot dependency management and will be replaced by Kotlin BOM when it will be available (see [KT-18398](https://youtrack.jetbrains.com/issue/KT-18398)).

`runApplication<FooApplication>()` is a form usually preferred in Kotlin to `runApplication(FooApplication::class)` and consistent with what we do on Spring Framework side. 

The custom Maven configuration is the [typical way to deal with Java + Kotlin project](https://kotlinlang.org/docs/reference/using-maven.html#compiling-kotlin-and-java-sources), a similar configuration has been applied to [Spring Data Kay](https://github.com/spring-projects/spring-data-build/blob/beb9d5cfdfdd18b0055db7648f859b92b0a168e5/parent/pom.xml#L775).

Closes gh-8511

sdeleuze commented:
The build failed for an unrelated flacky WebFlux test, but `./mvnw clean install` works fine locally and I have validated the top level function is usable in my MiXiT sample application.

sureshg commented on spring-boot-dependencies/pom.xml:
```diff
@@ -2285,6 +2286,26 @@
+			</dependency>
+			<dependency>
+				<groupId>org.jetbrains.kotlin</groupId>
+				<artifactId>kotlin-stdlib</artifactId>
```
Doesn't `kotlin-stdlib-jre8` transitively pull `kotlin-stdlib` and `kotlin-stdlib-jre7`?

snicoll commented on spring-boot-dependencies/pom.xml:
```diff
@@ -2285,6 +2286,26 @@
+			</dependency>
+			<dependency>
+				<groupId>org.jetbrains.kotlin</groupId>
+				<artifactId>kotlin-stdlib-jre7</artifactId>
```
We don't support Java 7 so that should go away.

wilkinsona replied:
I'm not sure I agree with that. I'd rather that the dependency management was complete.

snicoll replied:
I am not following. When are we going to use that dependency if we don't support Java 7 anymore?

sdeleuze replied:
We support Java 8 so kotlin-stdlib, kotlin-stdlib-jre7 and kotlin-stdlib-jre8 should be provided for consistent versioning. kotlin-stdlib-jre7 contains Java 7 specific methods which are needed for Java 8 support, same for kotlin-stdlib-jre8 and Java 8.

snicoll commented on spring-boot/pom.xml:
```diff
@@ -18,6 +18,79 @@
+					<parameters>true</parameters>
+				</configuration>
+				<executions>
+					<execution>
```
What are those? Can you explain?

snicoll replied:
Scratch that. I didn't read your description properly, sorry.

snicoll commented on spring-boot/src/test/kotlin/org/springframework/boot/SpringApplicationExtensionsTests.kt:
```diff
@@ -0,0 +1,73 @@
+
+		@Bean
+		open fun webServer(): TomcatServletWebServerFactory {
+			return TomcatServletWebServerFactory(0)
```
Is there a reason why you need to create a web server for this test? That looks a bit overkill to me.

sdeleuze replied:
Yes, because if not the test fails, I just copied what you do in SpringApplicationTests on Java side.

sdeleuze commented:
For Maven plugin configuration, it is the standard way to compile Java + Kotlin projects, I have on purpose only applied it to the module where we need that, we do the same in Spring Data without issues reported, and many other project do that. If you face issue, I can ask Kotlin team to fixit, if not I would recommand go that way since I have no alternative to propose.

sdeleuze commented:
Hum, there is already a `kotlin.run` top level function, so reusing that name for Boot may cause confusion. I will update the PR with another function name.

sdeleuze commented:
I have splitted this PR in 2 commits, one for dependency management, the other for the extension itself. I have also renamed the top level function to `runApplication()` which seems to me a good trade off between `run()` which will cause confusion with `kotlin.run()` and looks reasonably short (`runSpringApplication()` looks ugly for such widely used function).

sdeleuze commented:
I have also rebased this PR on top of the recent module structure changes.

GreyTeardrop commented:
Hi! Awesome to see even more Kotlin niceties in Boot!

@sdeleuze,
Do you think it could make sense to also add something like this:

```kotlin
inline fun <reified T : Any> runApplication(vararg args: String,
                                            block: SpringApplication.() -> Unit): ConfigurableApplicationContext =
    SpringApplication(T::class.java).apply(block).run(*args)
```

for cases when some `SpringApplication` tuning is required?

sdeleuze commented:
@GreyTeardrop Great idea, this makes this feature much more flexible.

@snicoll I have updated the PR with this proposal and added related tests. I have also created the related https://github.com/spring-projects/spring-boot/issues/10712 issue.

sdeleuze commented:
I have added plugin management for Maven plugin, after discussion with @wilkinsona sadly we can't do similar things with Gradle plugin. Unfortunate but IMO we can live with that, it is possible to have different minor versions between dependencies and plugin.

snicoll commented:
@wilkinsona this would require a Spring Boot version to Kotlin version mapping in start.spring.io for Gradle projecs and I am really not keen to add that. 

Would it be possible to extrac the `kotlin.version` from the BOM and use that as a property to define the version of the plugin to use? Or do something of the kind in our own dependency management plugin?

wilkinsona commented:
I'd have to investigate, but right now I don't know how to do it, particularly with the `plugins` DSL. We can only extract the `kotlin.version` from the bom once the Spring Boot and Dependency Management plugins have been applied. By that point, the Kotlin plugin may already have been applied too. I think there's a chicken and egg problem here.

wilkinsona commented:
We avoid the problem for the Dependency Management plugin by making the Spring Boot plugin pull it in transitively. That lets us control the version of the Dependency Management plugin and allows a user to just apply the plugin without worrying about its version. We could do the same for the Kotlin plugin, but then everyone building a Gradle project would download the Kotlin plugin even if they weren't using it. That's similar to what we did for Jackson's Kotlin module so we could decide that it's a reasonable thing to do, probably depending on the weight of the Kotlin plugin and its dependencies.

philwebb commented:
I'm guessing 90% of applications are going to do something like this:

```kotlin
fun main(args : Array<String>) { 
    runApplication<ExampleApp>(args)
}
```

I'm wondering if we shouldn't drop the multiple config version? Doing this:

```kotlin
runSpringApplication(arrayOf(ExampleConfig::class, ExampleWebConfig::class), args)
```

Doesn't seem much better than:

```kotlin
SpringApplication(ExampleConfig::class, ExampleWebConfig::class).run(args)
```

@sdeleuze Are there any benefits to the `arrayOf` version that I'm missing?

sdeleuze commented:
@philwebb Good point, indeed I think we can remove the `Array` variant.

philwebb commented:
I'll take care of that
